src/libraries/Common/src/SourceGenerators/CSharpSyntaxUtilities.cs

```diff
@@ -0,0 +1,68 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Globalization;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace SourceGenerators;
+
+internal static class CSharpSyntaxUtilities
+{
+    // Standard format for double and single on non-inbox frameworks.
+    public const string DoubleFormatString = "G17";
+    public const string SingleFormatString = "G9";
+
+    // Format a literal in C# format -- works around https://github.com/dotnet/roslyn/issues/58705
+    public static string FormatLiteral(object? value, TypeRef type)
```

Comment:
>FormatLiteral [](http://example.com/codeflow?start=25&length=13)

do we want to support Half type too which has precision of 5?

Reply:
I was basing these cases of what Roslyn special cased here: 
https://github.com/dotnet/roslyn/blob/0caa880a161057559d1dcaad890e4053f520136e/src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs#L494-L513

We can test `half` and see how it works.

Reply:
As best I can tell this is a gap in the language.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#6454-real-literals

Half literals are not supported.  We cannot emit a value for `Half` here as one doesn't exist today:
https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTBtAsgTwEEAXYgJymAFdiBTACnoAkIAbAMwEoBmAOhQAsnALqYE3AATIJAYQkBvTBOUSAxgHsYAZ2ISWHNQH0AFhIC8EvoIDcSleKkCJuZm3YTTF1xx78hCiQBfO2UHBCdcFG93Twlo30EuAOCMFSlJcOckaI9zSz8k+SCQ9MdnbhzYq38iwLTMFMwxDKRnIlIKajoJEAkSckoaWkxFVPtJAn7Oocq3ThLRtLSmPOM3WzHlFPrN0v13FaKAc1pia2KsDECgA

I filed https://github.com/dotnet/roslyn/issues/70300 track this gap in the language.